system-linux.c

```diff
@@ -1691,9 +1691,87 @@ int system_vlandev_del(struct device *vlandev)
 	return system_link_del(vlandev->ifname);
 }
 
+static inline int ethtool_link_mode_test_bit(unsigned int nr, const __u32 *mask)
+{
+	return !!(mask[nr / 32] & (1 << (nr % 32)));
+}
+
+static inline int ethtool_link_mode_set_bit(unsigned int nr, __u32 *mask)
+{
+	mask[nr / 32] |= (1 << (nr % 32));
+	return 0;
+}
+
+static int
+system_set_ethtool_settings_future(struct device *dev, struct device_settings *s)
+{
+	struct {
+		struct ethtool_link_settings req;
+		__u32 link_mode_data[3 * 127];
```

Comment:
Is there any reason, why are you inventing this custom struct content, magic numbers instead of copy&pasting the struct from the kernel https://github.com/torvalds/linux/blob/fcd476ea6a888ef6e6627f4c21a2ea8cca3e9312/net/ethtool/ioctl.c#L428 ?

Reply:
Well, you can't do that. The number of words used for the three mode bitfields has to be queried from the kernel at runtime and hence also determines the starting offsets of two of the three fields. Hence the best you can do is allocate enough space (`3 * INT8_MAX * sizeof(__u32)`) and then do some pointer juggling once `nwords` has been returned from the first call to `ETHTOOL_GLINKSETTINGS`.